capstone-backend/src/main/java/com/google/sps/servlets/Utility.java

```diff
@@ -58,4 +65,181 @@ public static JsonObject createRequestBodyJson(HttpServletRequest request) {
     }
     return jsonObject;
   }
+
+  /**
+   * Query the given collection by id. 
+   * @param {collectionReference} reference to the appropriate database collection
+   * @param {request} request sent to the backend
+   * @param {response} response returned from the call
+   * @param {genericClass} a Generic class, used in casting documents from the database
+   * @return List<T> a singleton List of the Object that matches the ID in the request.
+   */
+  private static <T> List<T> getById(CollectionReference collectionReference, HttpServletRequest request, 
+      HttpServletResponse response, GenericClass<T> genericClass) throws IOException {
+    if (request.getParameterMap().size() > 1) {
+      System.err.println("Error: No other parameter can be sent with an ID");
+      response.sendError(HttpServletResponse.SC_BAD_REQUEST);
+      return null;
+    }
+    List<T> retrievedObjects = new ArrayList<>();
+    ApiFuture<QuerySnapshot> asyncQuery;
+    DocumentReference docRef = collectionReference.document(request.getParameter("id"));
+    ApiFuture<DocumentSnapshot> asyncDocument = docRef.get();
+    DocumentSnapshot document = null;
+    T item = null;
+    try {
+      document = asyncDocument.get();
+      if (document.exists()) {
+        item = document.toObject(genericClass.getMyType());
+      } else {
+        System.err.println("Error: no such document!");
+        response.sendError(HttpServletResponse.SC_NOT_FOUND);
+        return null;
+      }
+    } catch (Exception e) {
+      System.err.println("Error: " + e);
+      response.sendError(HttpServletResponse.SC_INTERNAL_SERVER_ERROR);
+      return null;
+    }
+    retrievedObjects.add(item);
+    return retrievedObjects;
+  }
+
+  /**
+   * Appends filters to the given query depending on fields in the request query 
+   * @param {genericClass} a Generic class, used in casting documents from the database
+   * @param {response} response returned from the call
+   * @param {it} iterator for the Map of query parameter names to values
+   * @param {query} query being modified, to eventually be used to query the database.
+   * @return Query the query being modified, to eventually be used to query the database.
+   */
+  private static <T> Query addToQuery(GenericClass<T> genericClass, HttpServletResponse response, 
+      Iterator it, Query query) throws IOException {
+    Map.Entry<String, String[]> entry = (Map.Entry) it.next();
+    String parameterName = entry.getKey();
+    if (entry.getValue().length != 1) {
```

Comment:
Gitpod gives us a me a warning about ```(Map.Entry) it.next()``` 

Type safety: The expression of type Map.Entry needs unchecked conversion to conform to Map.Entry<String,String[]>

It also gives me a warning about ```Iterator```

Iterator is a raw type. References to generic type Iterator<E> should be parameterized

Reply:
I'll parameterize the iterator, good catch! Now that I've parameterized the iterator, I can remove the forced cast to type Map.Entry and that should deal with the Type Safety warning. Thanks for the comment!